### PR TITLE
[design doc] XGBoost on SQLFlow

### PR DESCRIPTION
**Pull request by Yancey0623**
A part work of #749

### CODE REVIEW

**Comment by Yancey0623** [comment]
Need to merge #754 first to keep the commit history of ant-xgboost design.

File: doc/xgboost_on_sqlflow_design.md [hunk 1]
```diff
@@ -1,152 +1,109 @@
...
+This design doc introduces how do users train/predict the [XGBoost](https://xgboost.ai/) model by SQLFlow SQL and how
+we implement it.
+
+## Design
+
+We prefer users to execute the SQLFlow Train/Predict SQL as follows:
+
+  ``` sql
+  SELECT * FROM train_table
+  TRAIN XGBoost
```
Comment (typhoonzero) [inline doc/xgboost_on_sqlflow_design.md]: `TRAIN XGBoost.someModel`?

File: doc/xgboost_on_sqlflow_design.md [hunk 2]
```diff
@@ -1,152 +1,109 @@
...
+  
+  ``` sql
+  SELECT * FROM test_table
+  PREDICT pred_table.result
+  USING my_xgb_model;
+  ```
+
+where:
+- `my_xgb_model` is the trained model.
+- The keyword `XGBOOST` is used to distinguish with the Tensorflow Model.
```
Comment (typhoonzero) [inline doc/xgboost_on_sqlflow_design.md]: did not see the `XGBOOST` keyword?

File: doc/xgboost_on_sqlflow_design.md [hunk 3]
```diff
@@ -1,152 +1,109 @@
...
+
+    ``` text
+    col1 | col2| col3 | label | group
+    1.1 2.0 2.2 1 1
+    0.8 2.0 2.2 2 1
+    0.2 3.0 4.2 0 2
+    0.77 4.0 2.6 2 3
+    ```
+
+    `codegen_xgboost.go` would write down the `train.txt.group` file like:
```
Comment (typhoonzero) [inline doc/xgboost_on_sqlflow_design.md]: Why need to write a file?

File: doc/xgboost_on_sqlflow_design.md (hunk 2 above)
Comment (Yancey0623) [reply doc/xgboost_on_sqlflow_design.md]: fixed it.

File: doc/xgboost_on_sqlflow_design.md (hunk 1 above)
Comment (Yancey0623) [reply doc/xgboost_on_sqlflow_design.md]: XGBoost use the `objective` paramter to specify the training objective such as:
`objective=binary:logistic `, ref https://xgboost.readthedocs.io/en/latest/parameter.html#learning-task-parameters

Maybe we can use `Train XGBoost` in the train statement, and specify the objective in WITH statement: `model.objective=binary:logistic` ?

File: doc/xgboost_on_sqlflow_design.md (hunk 3 above)
Comment (Yancey0623) [reply doc/xgboost_on_sqlflow_design.md]: XGboost use DMatrix as the input dataset , and it seems the text file  format is popular in XGBoost:
 
> XGBoost currently supports two text formats for ingesting data: LibSVM and CSV

ref: https://xgboost.readthedocs.io/en/latest/tutorials/input_format.html

File: doc/xgboost_on_sqlflow_design.md (hunk 1 above)
Comment (typhoonzero) [reply doc/xgboost_on_sqlflow_design.md]: I still prefer to put the `objective` in the `TRAIN` clause, it seems quite similar to `tf.estimator.*`.

Comment (Yancey0623) [reply doc/xgboost_on_sqlflow_design.md]: Update the doc, would use `TRAIN xgboost.multi.softmax` to fill the objective parameter.

File: doc/xgboost_on_sqlflow_design.md [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+
+## Design
+
+We prefer users to execute the SQLFlow Train/Predict SQL as follows:
+
+  ``` sql
+  SELECT * FROM train_table
+  TRAIN xgboost.multi.softmax
+  WITH
+      train.objective="multi:softmax",
```
Comment (typhoonzero) [inline doc/xgboost_on_sqlflow_design.md]: should remove `train.objective` now?

**Review by typhoonzero — DISMISSED** [review]
LGTM, just one line needs to be fixed

**Comment by Yancey0623** [comment]
Hi @wangkuiyi , thanks for correcting the grammar, I updated the design and remove some detail paragraphs:
1. remove the `Input Data Format` paragraph as we would support the group/weight instances in the future.
1. remove the  `Learning API` or `Scikit-Learn API` paragraph as it's too detailed.

File: doc/xgboost_on_sqlflow_design.md [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+    - the prefix `xgboost.` tells the model is a XGBoost one, but not a Tensorflow model, and
+    - `multi.softmax` names an [XGBoost learning task](https://xgboost.readthedocs.io/en/latest/parameter.html#learning-task-parameters).
+- In the `WITH` clause, 
+  - keys with the prefix `train.` identifies parameters of XGBoost API [`xgboost.train`](https://xgboost.readthedocs.io/en/latest/python/python_api.html#xgboost.train), and
+  - keys without any prefix identifies [XGBoost Parameters](https://xgboost.readthedocs.io/en/latest/parameter.html) except the `objective` parameter, which was specified by the identifier after the keyword `TRAIN`, as explained above.
+
+## The Code Generator
+
+The code generator `codegen_xgboost.go` outputs an XGBoost program in Python. It contains the following features:
+1. Execute the user-typed **SELECT STATEMENT** to retrieve the training data from SQL engine, then convert it to
```
Comment (wangkuiyi) [inline doc/xgboost_on_sqlflow_design.md]: **SELECT STATEMENT** => the SELECT statment

File: doc/xgboost_on_sqlflow_design.md [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+[XGBoost DMatrix](https://xgboost.readthedocs.io/en/latest/python/python_api.html?highlight=dmatrix#xgboost.DMatrix)
+which is the Data Matrix used inn XGBoost.
```
Comment (wangkuiyi) [inline doc/xgboost_on_sqlflow_design.md]: It tells the SQL engine to run the SELECT statement and retrieve the training/test data.  It saves the data into a text file, which could be loaded by XGBoost using the DMatrix interface.

File: doc/xgboost_on_sqlflow_design.md [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+1. Parse and resolve the **WITH** clause to fill the `xgboost.train` arguments and the XGBoost Parameters.
+1. Save the trained model on disk.
+1. For the **PREDICT STATEMENT**, the submitter Python program would load the trained model and test data to output the prediction result to a SQL engine.
```
Comment (wangkuiyi) [inline doc/xgboost_on_sqlflow_design.md]: PREDICT clause

File: doc/xgboost_on_sqlflow_design.md (hunk 5 above)
Comment (Yancey0623) [reply doc/xgboost_on_sqlflow_design.md]: done.

File: doc/xgboost_on_sqlflow_design.md (hunk 6 above)
Comment (Yancey0623) [reply doc/xgboost_on_sqlflow_design.md]: Thanks, done.

File: doc/xgboost_on_sqlflow_design.md (hunk 7 above)
Comment (Yancey0623) [reply doc/xgboost_on_sqlflow_design.md]: Done, and also polish the last sentence.

File: doc/xgboost_on_sqlflow_design.md [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+    - `multi.softmax` names an [XGBoost learning task](https://xgboost.readthedocs.io/en/latest/parameter.html#learning-task-parameters).
+- In the `WITH` clause, 
+  - keys with the prefix `train.` identifies parameters of XGBoost API [`xgboost.train`](https://xgboost.readthedocs.io/en/latest/python/python_api.html#xgboost.train), and
+  - keys without any prefix identifies [XGBoost Parameters](https://xgboost.readthedocs.io/en/latest/parameter.html) except the `objective` parameter, which was specified by the identifier after the keyword `TRAIN`, as explained above.
+
+## The Code Generator
+
+The code generator `codegen_xgboost.go` outputs an XGBoost program in Python. It contains the following features:
+1. It tells the SQL engine to run the SELECT statement and retrieve the training/test data. It saves the data into a text file, which could be loaded by XGBoost using the DMatrix interface.
+1. Parse and resolve the WITH clause to fill the `xgboost.train` arguments and the XGBoost Parameters.
```
Comment (wangkuiyi) [inline doc/xgboost_on_sqlflow_design.md:43]: I think the parsing of the WITH clause is the parser's work, but not the submitter's work, am I right?

**Review by wangkuiyi — APPROVED** [review]

File: doc/xgboost_on_sqlflow_design.md (hunk 8 above)
Comment (Yancey0623) [reply doc/xgboost_on_sqlflow_design.md:43]: The parser can parse the WITH clause to a general [attrs](https://github.com/sql-machine-learning/sqlflow/blob/develop/sql/parser.go#L106) struct which is a Go struct `map[string]*expr`, and each generator would resolve the`attrs` to program parameters, such as XGBoost generator would convert the attrs as follows:
- keys with `train.` prefix to `xgboost.train` arguments.
- keys without any prefix to XGBoost Parameters which is JSON format.

File: doc/xgboost_on_sqlflow_design.md (hunk 1 above)
Comment (tonyyang-svail) [reply doc/xgboost_on_sqlflow_design.md]: > I still prefer to put the `objective` in the `TRAIN` clause, it seems quite similar to `tf.estimator.*`.

@Yancey1989 @typhoonzero `objective` corresponds to the loss of function of a model. So it shouldn't be in the model name. For different types of models, there are `gbtree`, `gblinear` and `dart` as listed [here](https://xgboost.readthedocs.io/en/latest/parameter.html#general-parameters).

Comment (typhoonzero) [reply doc/xgboost_on_sqlflow_design.md]: @tonyyang-svail Thanks, you are right. I'll update to put`objective` in attributes.
